llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -479,6 +479,12 @@ RISCVTargetLowering::RISCVTargetLowering(const TargetMachine &TM,
     // FIXME: Need to promote bf16 FCOPYSIGN to f32, but the
     // DAGCombiner::visitFP_ROUND probably needs improvements first.
     setOperationAction(ISD::FCOPYSIGN, MVT::bf16, Expand);
+
+    // To fold (bf16 bitcast (copyfromreg f16)) -> (copyfromreg bf16), we have
+    // to legalize the f16 CopyFromReg for avoiding SoftPromoteHalf.
+    setOperationAction(ISD::CopyFromReg, MVT::f16, Legal);
+    // Fold the (bf16 bitcast (copyfromreg f16)) -> (copyfromreg bf16).
+    setTargetDAGCombine(ISD::BITCAST);
```

Comment:
The bitcast handling shouldn't be strictly necessary for the asm support and should be done in a separate PR

Reply:
If I don't combine the bitcast this time, the new test I added would be fail since RISCV doesn't have the instruction for selecting the `bf16 bitcast fp16`. In the IR level, it even doesn't allow the bitcast between fp16 <--> bf16.

Reply:
Then this is entirely incorrect. You cannot rely on this DAG combine to fix up a selection failure

Reply:
OK, sorry for that I doesn't realize about this policy originally. I will seek the correct way to handle it, thanks for the comment!

Reply:
You probably need to add bf16 as a legal type if you want to support registers that directly contain it (and then directly handle bitcasts in selection). That's approximately what legal means anyway. You can only get so far by hacking it into making CopyFromReg legal

Reply:
Yes, bitcast between fp16 and bf16 is allowed, sorry for the misunderstanding.